cookbook/x-flat-jwt.md

```diff
@@ -0,0 +1,324 @@
+HTTP/1.1 403 Forbidden
+…
+Content-Type: application/json
+
+{"error":{"message":"Security violation","status":403,"requestID":"Xsz80Av8CEBZfWbmQhKLIAAAAIE","info":["JWT Security (JWTCookie): Invalid JWT: Token has expired."],"code":3206}}
+```
+
+Hmm, expired. So this one is too old. (Access tokens typically have a restricted period of use.)
+
+OK, let's use another token:
+
+```
+curl -i -H "Cookie: authtoken=eybGciOiJSUzI1NiJ9.eyJzdWIiOiJzb21lX3VzZXIiLCJpc3MiOiJzb21lX3Byb3ZpZGVyIn0.bNXv28XmnFBjirPbCzBqyfpqHKo6PpoFORHsQ-80IJLi3IhBh1y0pFR0wm-2hiz_F7PkGQLTsnFiSXxCt1DZvMstbQeklZIh7O3tQGJyCAi-HRVASHKKYqZ_-eqQQhNr8Ex00qqJWD9BsWVJr7Q526Gua7ghcttmVgTYrfSNDzU" http://localhost:1234/httpbin/anything
+HTTP/1.1 200 OK
+…
+Content-Type: application/json
+
+{
+  "args": {},
+  "data": "",
+  "files": {},
+  "form": {},
+  "headers": {
+    "Accept": "*/*",
+    "Accept-Encoding": "deflate, gzip",
+    "Host": "httpbin.org",
+    "User-Agent": "curl/7.29.0",
+    "X-Amzn-Trace-Id": "…" 
+  },
+  "json": null,
+  "method": "GET",
+  "origin": "…",
+  "url": "https://httpbin.org/anything" 
+}
+```
+
+Tada!
+
+By the way, apart from Cookies, this also works similarly with the `Authorization: Bearer …` header:
+
+```
+…
+securityDefinitions:
+  JWTBearer:
+    type: apiKey
+    in: header
+    name: Authorization             # ⬅
+    x-flat-jwt:
+      key:
+        file: secret.pem
+      alg: RS256
+paths:
+  /httpbin/**:
+    security:
+      - JWTBearer: []
+```
+
+You can try that with
+
+```
+curl -i -H "Authorization: Bearer eybGciOiJSUzI1NiJ9.eyJzdWIiOiJzb21lX3VzZXIiLCJpc3MiOiJzb21lX3Byb3ZpZGVyIn0.bNXv28XmnFBjirPbCzBqyfpqHKo6PpoFORHsQ-80IJLi3IhBh1y0pFR0wm-2hiz_F7PkGQLTsnFiSXxCt1DZvMstbQeklZIh7O3tQGJyCAi-HRVASHKKYqZ_-eqQQhNr8Ex00qqJWD9BsWVJr7Q526Gua7ghcttmVgTYrfSNDzU" http://localhost:1234/httpbin/anything
+```
+
+But there are two additional features that can be quite handy: `out-var` and `out-header`.
+
+## Storing JWT claims: `out-var`
+
+With `out-var` you can specify the name of a [variable](/reference/variables.md) where FLAT will store the JSON claims encoded in the JWT, in order to make them available for further processing. E.g.
+
+
+```yaml
+…
+    x-flat-jwt:
+      key:
+        file: secret.pem
+      alg: RS256
+      out-var: $the_claims          # ⬅
+…
+```
+
+We can log the claims by adding a `log` action to an [init flow](/reference/OpenAPI/routing.md#init-flow):
+
+```yaml
+swagger: "2.0" 
+…
+x-flat-init: init.xml
+paths:
+…
+```
+
+with init.xml:
+
+```xml
+<flow>
+  <log>
+  {
+    "JWT-Claims": {{ $the_claims }}
+  }
+  </log>
+</flow>
+```
+
+If you look at the [FLAT logs](/administration/logging.md) and try again with a valid token, you'll notice:
+
+```
+{…,"type":"flat_access",…,"JWT-Claims":{"sub":"some_user","iss":"some_provider"}}
+```
+
+Here you see the two claims from the JWT token.
+
+## Sending JWT claims upstream: `out-header`
```

Comment:
i would move this section up. the out-header plays well with proxy and is necessary in many situations (whereas logging is optional).